### PR TITLE
Data API as library - Offline Mode

### PR DESCRIPTION
**Pull request by kathirsvn**
**What this PR does**:
Changes to use Data API as a library in offline mode (i.e. not as a service). Below are the commands introduced as part of this change.
- beginOfflineSession
- offlineInsertMany
- offlineGetStatus
- endOfflineSession

**Which issue(s) this PR fixes**:
Fixes NA

**Checklist**
- [X] Changes manually tested
- [X] Automated Tests added/updated
- [X] Documentation added/updated
- [X] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -247,6 +247,101 @@
...
+          <snapshots>
+            <enabled>true</enabled>
+          </snapshots>
+        </repository>
+      </repositories>
+      <dependencies>
+        <dependency>
+          <groupId>com.datastax.dse</groupId>
+          <artifactId>dse-db-all</artifactId>
+          <version>4.0.11-591d171ac9c9</version>
```
Comment (tatu-at-datastax) [inline pom.xml:269]: We have this version referenced from:

```
<stargate.int-test.cassandra.image-tag>
```

but with different actual value -- should probably create a common property to reference (add in main `<properties>` so accessible to profile section too, I think)?

File: src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java [hunk 2]
```diff
@@ -19,6 +19,12 @@ public class DataApiRequestInfo {
   private final Optional<String> cassandraToken;
   private final Optional<String> embeddingApiKey;
 
+  public DataApiRequestInfo(Optional<String> tenantId) {
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java:22]: Maybe add javadoc indicating why this constructor needed? (for offline use)

File: src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCode.java [hunk 3]
```diff
@@ -157,6 +157,8 @@ public enum ErrorCode {
   UNAUTHENTICATED_REQUEST("UNAUTHENTICATED: Invalid token"),
   COLLECTION_CREATION_ERROR(
       "Collection creation failure (unable to create table). Recommend re-creating the collection"),
+  OFFLINE_WRITER_SESSION_NOT_FOUND("Offline writer session not found :"),
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCode.java:160]: No need for trailing ` :` -- when using `ErrorCode.xxx.toApiException()` separate gets added automatically.

File: src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/CQLSessionCache.java [hunk 4]
```diff
@@ -51,14 +51,17 @@ public class CQLSessionCache {
...
+  /** Persistence type SSTable Writer */
+  public static final String OFFLINE_WRITER = "offline_writer";
 
   @ConfigProperty(name = "quarkus.application.name")
-  private String APPLICATION_NAME;
+  String APPLICATION_NAME;
 
   @Inject
   public CQLSessionCache(OperationsConfig operationsConfig, MeterRegistry meterRegistry) {
+    LOGGER.info("Initializing CQLSessionCache");
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/CQLSessionCache.java:62]: Is this generally useful? Or only during development -- has very little information.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/InsertOperation.java [hunk 5]
```diff
@@ -159,28 +170,30 @@ private static Uni<DocumentId> insertDocument(
...
-  private String buildInsertQuery(boolean vectorEnabled) {
+  public String buildInsertQuery(boolean vectorEnabled) {
     if (vectorEnabled) {
       String insertWithVector =
           "INSERT INTO \"%s\".\"%s\""
               + " (key, tx_id, doc_json, exist_keys, array_size, array_contains, query_bool_values, query_dbl_values , query_text_values, query_null_values, query_timestamp_values, query_vector_value)"
               + " VALUES"
-              + " (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)  IF NOT EXISTS";
+              + " (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) "
+              + (offlineMode ? "" : " IF NOT EXISTS");
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/InsertOperation.java:180]: Should be explained somewhere why "IF NOT EXISTS" not used in offlineMode?
(not needed as we are creating new sstables and not modifying existing tables?)

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/InsertOperation.java [hunk 6]
```diff
@@ -195,7 +208,9 @@ private static SimpleStatement bindInsertValues(
           CQLBindValues.getDoubleMapValues(doc.queryNumberValues()),
           CQLBindValues.getStringMapValues(doc.queryTextValues()),
           CQLBindValues.getSetValue(doc.queryNullValues()),
-          CQLBindValues.getTimestampMapValues(doc.queryTimestampValues()),
+          offlineMode
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/InsertOperation.java:211]: Should have a note on why difference needed? (cannot unify for some reason)

File: src/main/offline/io/stargate/sgv2/jsonapi/api/request/FileWriterParams.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        || tableName.isBlank()
+        || ssTableOutputDirectory == null
+        || ssTableOutputDirectory.isBlank()
+        || fileWriterBufferSizeInMB <= 0
+        || createTableCQL == null
+        || createTableCQL.isBlank()
+        || insertStatementCQL == null
+        || insertStatementCQL.isBlank()
+        || vectorEnabled == null) {
+      throw new IllegalArgumentException("Invalid FileWriterParams");
```
Comment (tatu-at-datastax) [inline src/main/offline/io/stargate/sgv2/jsonapi/api/request/FileWriterParams.java:27]: Ok for now but probably need to change in near future to indicate exactly which argument was bad.

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/CommandContext.java [hunk 8]
```diff
@@ -25,6 +25,16 @@ public CommandContext(String namespace, String collection) {
     this(namespace, collection, CollectionSettings.empty(), null, null, null);
   }
 
+  public static CommandContext from(
```
Comment (maheshrajamani) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/CommandContext.java:28]: java doc

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/CommandStatus.java [hunk 9]
```diff
@@ -47,5 +47,18 @@ public enum CommandStatus {
...
    */
   @JsonProperty("upsertedId")
-  UPSERTED_ID;
+  UPSERTED_ID,
+  /**
+   * The element has the session id of offline writer, which is used to write the data offline to
+   * the SSTable files for example
+   */
+  @JsonProperty("offlineWriterSessionId")
+  OFFLINE_WRITER_SESSION_ID,
```
Comment (maheshrajamani) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/CommandStatus.java:56]: Not a big deal but this file has been maintained in alphabetic order.

File: src/main/java/io/stargate/sgv2/jsonapi/exception/JsonApiException.java [hunk 10]
```diff
@@ -63,7 +65,12 @@ public CommandResult get() {
   public CommandResult.Error getCommandResultError(String message, Response.Status status) {
     Map<String, Object> fieldsForMetricsTag =
         Map.of("errorCode", errorCode.name(), "exceptionClass", this.getClass().getSimpleName());
-    SmallRyeConfig config = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class);
+    SmallRyeConfig config;
```
Comment (maheshrajamani) [inline src/main/java/io/stargate/sgv2/jsonapi/exception/JsonApiException.java:68]: Can you add comment on why this is done. Can the `config` made static

File: src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineCommandsProcessor.java [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+  private OperationsConfig operationsConfig;
+  private CQLSessionCache cqlSessionCache;
+  private CommandResolverService commandResolverService;
+  private DataVectorizerService dataVectorizerService;
+  private static OfflineCommandsProcessor offlineCommandsProcessor;
+  private static boolean initialized;
+
+  private OfflineCommandsProcessor() {}
+
+  public static OfflineCommandsProcessor getInstance() {
```
Comment (tatu-at-datastax) [inline src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineCommandsProcessor.java:49]: Instead of separate `init()`, could just have

```
   private final OfflineCommandsProcessor offlineCommandsProcessor = new OfflineCommandsProcessor();

   public static OfflineCommandsProcessor getInstance() { return offlineCommandsProcessor; }
```

and do everything `init()` has now in constructor? That would change most fields to be `final` as well.

File: src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineFileWriterInitializer.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+import io.smallrye.config.SmallRyeConfig;
+import io.smallrye.config.SmallRyeConfigBuilder;
+import io.smallrye.context.SmallRyeContextManagerProvider;
+import io.stargate.sgv2.jsonapi.config.OperationsConfig;
+import io.stargate.sgv2.jsonapi.config.constants.ApiConstants;
+
+public class OfflineFileWriterInitializer {
+
+  public static void initialize() {
+    System.setProperty(ApiConstants.OFFLINE_WRITER_MODE_PROPERTY, "true");
```
Comment (tatu-at-datastax) [inline src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineFileWriterInitializer.java:14]: This seems.... somehow wrong. System properties should not usually be changed by running Java code. Especially when there's timing question (something having to call initialize()).
Is there no other way to achieve this?

File: src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineFileWriterInitializer.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    SmallRyeContextManagerProvider.instance();
```
Comment (tatu-at-datastax) [inline src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineFileWriterInitializer.java:15]: Why is this needed?

File: src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineWriterSessionStatus.java [hunk 14]
```diff
@@ -0,0 +1,11 @@
+package io.stargate.sgv2.jsonapi.service.cqldriver.sstablewriter;
+
+public record OfflineWriterSessionStatus(
+    String sessionId,
+    String keyspace,
+    String tableName,
+    String ssTableOutputDirectory,
+    int fileWriterBufferSizeInMB,
+    int dataDirectorySizeInBytes,
```
Comment (tatu-at-datastax) [inline src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineWriterSessionStatus.java:9]: Should this be `long` instead? (max of `int` is about 2 gigs)

File: src/main/offline/io/stargate/sgv2/jsonapi/service/operation/model/impl/BeginOfflineSessionOperation.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+      FileWriterSession fileWriterSession =
+          new FileWriterSession(
+              queryExecutor.getCqlSessionCache(), sessionCacheKey, sessionId, fileWriterParams);
+      queryExecutor.getCqlSessionCache().putSession(sessionCacheKey, fileWriterSession);
+      CommandResult commandResult =
+          new CommandResult(Map.of(CommandStatus.OFFLINE_WRITER_SESSION_ID, sessionId));
+      return Uni.createFrom().item(() -> () -> commandResult);
+    } catch (Exception e) {
+      JsonApiException jsonApiException =
+          new JsonApiException(
```
Comment (tatu-at-datastax) [inline src/main/offline/io/stargate/sgv2/jsonapi/service/operation/model/impl/BeginOfflineSessionOperation.java:34]: There is more convenient mechanism with `ErrorCode.xxx.toApiException()`.

**Review by maheshrajamani — COMMENTED** [review]
Provided few comments, In general many new classes need java doc to be added.
Can you add a ReadMe file for how to use offline mode?

**Review by tatu-at-datastax — APPROVED** [review]
Looks good, added notes -- but I am hoping someone else (Mahesh?) can also review.

File: pom.xml (hunk 1 above)
Comment (kathirsvn) [reply pom.xml]: Fixed in this commit [4167d29](https://github.com/stargate/data-api/pull/905/commits/4167d29fe3f3d950025e65d9b06eab947a619088)

File: src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java (hunk 2 above)
Comment (kathirsvn) [reply src/main/java/io/stargate/sgv2/jsonapi/api/request/DataApiRequestInfo.java:28]: Updated in this commit [3bf31d2](https://github.com/stargate/data-api/pull/905/commits/3bf31d24b56730eb75c36b05152633df26111ef9)

File: src/main/java/io/stargate/sgv2/jsonapi/config/constants/ApiConstants.java [hunk 16]
```diff
@@ -0,0 +1,9 @@
+package io.stargate.sgv2.jsonapi.config.constants;
+
+public class ApiConstants {
```
[the remark this answers is not in the document]
Comment (kathirsvn) [reply src/main/java/io/stargate/sgv2/jsonapi/config/constants/ApiConstants.java:3]: I agree this is being separate from the Quarkus configurations that we maintain. But this is used in a place where the config property of a specific config group is retrieved from `SmallRyeConfig` instance and parsing this property itself from a `Smallrye` config is not possible.
```java
if (ApiConstants.isOffline()) {
  config = new SmallRyeConfigBuilder().withMapping(DebugModeConfig.class).build();
} else {
  config = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class);
}
```

[the remark this answers is not in the document]
Comment (kathirsvn) [reply src/main/java/io/stargate/sgv2/jsonapi/config/constants/ApiConstants.java:10]: Updated javadoc in this commit [882d1dc](https://github.com/stargate/data-api/pull/905/commits/882d1dcf853dc34e0e7935851d193ac7eee2d7b2)

File: src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCode.java (hunk 3 above)
Comment (kathirsvn) [reply src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCode.java]: Fixed in this commit [40402ef](https://github.com/stargate/data-api/pull/905/commits/40402efce668cca5834c2acc9fbc65ebebe1a8aa)

File: src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/CQLSessionCache.java (hunk 4 above)
Comment (kathirsvn) [reply src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/CQLSessionCache.java:62]: This is generally useful to have a log at the beginning and at the end of this constructor which happens only once during the start up/when the first request is received.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/InsertOperation.java (hunk 5 above)
Comment (kathirsvn) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/InsertOperation.java:180]: Added comment in this commit https://github.com/stargate/data-api/pull/905/commits/af4044312cc8c08c6351667c7c7b89a657fa00fd

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/InsertOperation.java (hunk 6 above)
Comment (kathirsvn) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/InsertOperation.java:217]: Added comment in this commit [9f2dc01](https://github.com/stargate/data-api/pull/905/commits/9f2dc01ab7a78ceedb05387a8c34f2febed0fb19)

File: src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineCommandsProcessor.java (hunk 11 above)
Comment (kathirsvn) [reply src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineCommandsProcessor.java:57]: I still have the `init` method for handling the synchronization. I'll test and remove that if classloader initialization takes care of it (I added a TODO marker there for now).

File: src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineFileWriterInitializer.java (hunk 13 above)
Comment (kathirsvn) [reply src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineFileWriterInitializer.java:15]: This is needed one time in the startup for the SmallRye config parsing to work in the subsequent flows

File: src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineFileWriterInitializer.java (hunk 12 above)
Comment (kathirsvn) [reply src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineFileWriterInitializer.java:14]: I agree with you on the timing part. Since, invoking this `initialize` method one time during start up is a 'must' for the offline sstable writer, I thought of leaving it here. Apart from this vm arg, invoking `instance()` method of `SmallRyeContextManagerProvider` is also required during start up and so I put them together in this startup method.

File: src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineWriterSessionStatus.java (hunk 14 above)
Comment (kathirsvn) [reply src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineWriterSessionStatus.java]: Good catch. Fixed in this commit [ae61b15](https://github.com/stargate/data-api/pull/905/commits/ae61b15f85a0e05c1273572db5e6b8ee6bab0163).

File: src/main/offline/io/stargate/sgv2/jsonapi/service/operation/model/impl/BeginOfflineSessionOperation.java (hunk 15 above)
Comment (kathirsvn) [reply src/main/offline/io/stargate/sgv2/jsonapi/service/operation/model/impl/BeginOfflineSessionOperation.java]: This is fixed in this commit [6204227](https://github.com/stargate/data-api/pull/905/commits/6204227107288e9f9ef3eb1758ffa423bbcf4cd7).

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/CommandContext.java (hunk 8 above)
Comment (kathirsvn) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/CommandContext.java:39]: Added in this commit [b403d2c](https://github.com/stargate/data-api/pull/905/commits/b403d2c798a6226f0738fa42e582b7329b49f304).

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/CommandStatus.java (hunk 9 above)
Comment (kathirsvn) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/CommandStatus.java:43]: Changes done in this commit [fb57ad3](https://github.com/stargate/data-api/pull/905/commits/fb57ad39c79f6eafb7878e785744a8bb5a2bfe07).

File: src/main/java/io/stargate/sgv2/jsonapi/exception/JsonApiException.java (hunk 10 above)
Comment (kathirsvn) [reply src/main/java/io/stargate/sgv2/jsonapi/exception/JsonApiException.java]: Done. Fixed in this commit [02bf4d0](https://github.com/stargate/data-api/pull/905/commits/02bf4d024b669e0e7ae70f09c804b2157646b501).

File: src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/executor/CollectionSettings.java [hunk 17]
```diff
@@ -306,7 +306,7 @@ public static CollectionSettings getCollectionSettings(
         collectionName, vectorEnabled, vectorSize, similarityFunction, comment, objectMapper);
   }
 
-  private static CollectionSettings createCollectionSettings(
+  public static CollectionSettings createCollectionSettings(
```
[the remark this answers is not in the document]
Comment (kathirsvn) [reply src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/executor/CollectionSettings.java]: Changed it back to `private` since it no longer needs to be public, which was required temporarily during one of the tests.

File: src/main/offline/io/stargate/sgv2/jsonapi/api/request/FileWriterParams.java [hunk 18]
```diff
@@ -0,0 +1,30 @@
+package io.stargate.sgv2.jsonapi.api.request;
+
+import java.util.List;
+
+public record FileWriterParams(
```
[the remark this answers is not in the document]
Comment (kathirsvn) [reply src/main/offline/io/stargate/sgv2/jsonapi/api/request/FileWriterParams.java:8]: Added in this commit [35b133a](https://github.com/stargate/data-api/pull/905/commits/35b133aff051eb5527c122f7520e92b135cad9f0)

File: src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/FileWriterSession.java [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.cql3.functions.types.CodecRegistry;
+import org.apache.cassandra.cql3.functions.types.DataType;
+import org.apache.cassandra.cql3.functions.types.TupleType;
+import org.apache.cassandra.cql3.functions.types.TupleValue;
+import org.apache.cassandra.io.sstable.CQLSSTableWriter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FileWriterSession implements CqlSession {
```
[the remark this answers is not in the document]
Comment (kathirsvn) [reply src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/FileWriterSession.java:51]: Added in [0ef06c8](https://github.com/stargate/data-api/pull/905/commits/0ef06c8c77586763d77880fe24b53d5004f06742)

File: src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineModeConstants.java [hunk 20]
```diff
@@ -0,0 +1,6 @@
+package io.stargate.sgv2.jsonapi.service.cqldriver.sstablewriter;
+
+public class OfflineModeConstants {
```
[the remark this answers is not in the document]
Comment (kathirsvn) [reply src/main/offline/io/stargate/sgv2/jsonapi/service/cqldriver/sstablewriter/OfflineModeConstants.java]: Fixed in this commit [43b4809](https://github.com/stargate/data-api/pull/905/commits/43b4809e4abf48c200742561ad57e78bcb0ba9fa)

File: src/main/java/io/stargate/sgv2/jsonapi/exception/JsonApiException.java (hunk 10 above)
Comment (kathirsvn) [reply src/main/java/io/stargate/sgv2/jsonapi/exception/JsonApiException.java:69]: @maheshrajamani reverted this commit, since it breaks the JsonApiException Initialization during application startup.
